src/uuid.cpp

```diff
@@ -0,0 +1,87 @@
+#include "uuid.h"
+
+#include <random>
+#include <limits>
+#include <sstream>
+#include <iomanip>
+
+static std::random_device rd;
+static std::mt19937 eng;
+
+/** use a distribution as mt19937 only guarantees minimum (not exactly) 32 bits */
+static std::uniform_int_distribution<> distrib( 0, std::numeric_limits<uint32_t>::max() );
+
+static inline void gen_uuid4( uint32_t *data ) {
+	data[ 0 ] = distrib( eng );
+	data[ 1 ] = distrib( eng );
+	data[ 2 ] = distrib( eng );
+	data[ 3 ] = distrib( eng );
+}
+
+void uuid::init() {
+	eng.seed( rd() );
+}
+
+uuid::uuid() {
+	gen_uuid4( data );
+}
+
+uuid::uuid( const uuid & ) {
```

Comment:
We have pseudocode like this:

1. Get item A
2. Create copy as item B
3. Manipulate both A and B
4. Drop both A and B to same location

Both A and B would have the duplicate `uuid` with the naive copy constructor.

I'd argue quite the opposite - we have to somehow make `uuid` unique and I'd rather the class enforced this rather than relying upon consumers to be careful